### PR TITLE
Web service report

### PR DESCRIPTION
**Pull request by katiasuya**

### CODE REVIEW

File: report.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	"strconv"
+
+	"github.com/katiasuya/parentheses/brackets"
+)
+
+// N is a number of times to run the service with each given length from Lens.
+const N = 1000
+
+// Lens array represents lengths of brackets sequences.
+var Lens = []int{2, 4, 6, 8, 10}
```
Comment (papahmsolo) [inline report.go:17]: Do we need this slice?

File: report.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+func main() {
+	var count int
+	for _, n := range Lens {
+		count = 0
+		for i := 0; i < N; i++ {
+
+			resp, err := http.Get("http://localhost:8000/generate?n=" + strconv.Itoa(n))
```
Comment (papahmsolo) [inline report.go:25]: Let's extract the URL into a constant

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

- Add info about how to generate a report
- We need to handle the situation when the response from the service was not successful.

File: README.md [hunk 3]
```diff
@@ -9,3 +9,7 @@ in the right order:
 
     func IsBalanced(str string) bool
 
+It also contains a report file that calculates the percentage of balanced sequences of brackets of different lengths 
+and prints the results to stdout.
+
+To generate a report, run the service and call it using GET method to /generate, passing a length n to it as a query parameter.
```
Comment (papahmsolo) [inline README.md:15]: How do I run the server and how to generate the report?

File: report.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	// URLGenerate is a URL for brackets generator service.
+	URLGenerate = "http://localhost:8000/generate?n="
+)
+
+func main() {
+	var count, length int
+
+	for length <= 10 {
+		count = 0
+		length += 2
```
Comment (papahmsolo) [inline report.go:25]: We can move `count` and `length` into the `for` statement

File: report.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+		for i := 0; i < N; i++ {
+			resp, err := http.Get(URLGenerate + strconv.Itoa(length))
```
Comment (papahmsolo) [inline report.go:28]: We're evaluating the same url N times

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Cool!

Let's not terminate the report upon the first error we encounter, log it and move on

File: report.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+const (
+	// N is a number of times to run the service with each length.
+	N = 1000
+	// URLGenerate is a URL for brackets generator service.
+	URLGenerate = "http://localhost:8000/generate?n="
+)
+
+func main() {
+	for length, count := 2, 0; length <= 10; length += 2 {
+		reportURL := URLGenerate + strconv.Itoa(length)
```
Comment (papahmsolo) [inline report.go:22]: You can use `fmt.Sprint(length)` to add `n` to URL or make `reportURL` a template and use `Sprintf`

File: report.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+				b, _ := ioutil.ReadAll(resp.Body)
+				log.Println(string(b))
+				return
+			}
+
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				log.Println(err)
+				return
+			}
```
Comment (papahmsolo) [inline report.go:42]: Looks like we're duplicating the `ioutil.ReadAll(resp.Body)`

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Very good!

Note, that you have two `main` functions in the same package in `report.go` and `main.go`

File: report/report.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+			body, err := ioutil.ReadAll(resp.Body)
+			if resp.StatusCode != http.StatusOK {
+				log.Println(string(body))
+				return
+			}
+			if err != nil {
+				log.Println(err)
+				return
+			}
```
Comment (papahmsolo) [inline report/report.go:39]: - Let's first check the error
- Let's not terminate the report if an error occurs, we can log it and consider the sequence as unbalanced

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Cool!

It makes sense to call the generate service concurrently as the requests are independent.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
There's a data race in your code, try running it with `-race` flag

File: report/report.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+					isBalanced <- true
+					return
+				}
+
+				isBalanced <- false
+			}()
+
+			if <-isBalanced {
+				count++
+			}
```
Comment (papahmsolo) [inline report/report.go:67]: This is waiting to receive from the channel, so it won't go to the next iteration until it receives from `isBalanced`

**Review by papahmsolo — CHANGES_REQUESTED** [review]

File: report/report.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New(string(body))
+	}
+
+	return body, nil
```
Comment (papahmsolo) [inline report/report.go:37]: You can return a string instead of []byte, which will save you a conversion.

File: report/report.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+				for u := range urls {
+					body, err := serviceResponse(u)
+					if err != nil {
+						log.Println(err)
+						isBalanced <- false
+						return
+					}
+					isBalanced <- brackets.IsBalanced(string(body))
+				}
+			}()
```
Comment (papahmsolo) [inline report/report.go:60]: So you're spinning up `numRequests` workers here, cool. Note, that if you `return` in this function the goroutine finishes execution, and the worker it's no longer a worker :)

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Cool, let's make some adjustments

File: report/report.go (hunk 11 above)
Comment (katiasuya) [reply report/report.go:60]: You're right!

**Review by papahmsolo — APPROVED** [review]
Good job

**Pull request merged**